source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp

```diff
@@ -0,0 +1,637 @@
+          // This moves ownership of the buffer to the new request in the new queue
+          // We do this so that we don't have lock any thread while waiting for the requests to complete
+          const rsl::unique_lock closed_req_lock(g_closed_request_mutex);
+          g_closed_requests.push_back(rsl::move(request));
+        }
+
+        using namespace rsl::chrono_literals; // NOLINT(google-build-using-namespace)
+        rsl::this_thread::sleep_for(1ms);
+      }
+    }
+
+    void wait_for_read_requests()
+    {
+      while (g_vfs_state_controller.has_state(VfsState::Running))
+      {
+        const rsl::unique_lock lock(g_closed_request_mutex);
+        if (!g_closed_requests.empty())
+        {
+          // Go over all the queued request and check if their read requests are all done
+          for (rsl::unique_ptr<QueuedRequest>& request : g_closed_requests)
+          {
+            if (request->all_requests_finished())
+            {
+              request.reset();
+            }
+          }
+
+          // Remove all requests that are nullptr, aka all requests that have finished
+          auto it = rsl::remove(g_closed_requests.begin(), g_closed_requests.end(), nullptr);
+          g_closed_requests.erase(it, g_closed_requests.end());
+        }
+
+        using namespace rsl::chrono_literals; // NOLINT(google-build-using-namespace)
+        rsl::this_thread::sleep_for(1ms);
+      }
+    }
+
+    void start_threads()
+    {
+      // This thread processes the read requests
+      // A queued request is popped from the queue
+      // reads the file specified as the requested file to read
+      // signals all the read requests that this file is read
+      // and passed in the buffer with its size
+      // After which it waits async for all read requests to have finished processing
+      // the data passed in to them
+      // Finally it passes itself over to the closing thread queue
+      // to make sure the data stays alive until all requests have finished
+      g_reading_thread = rsl::thread(process_read_requests);
+
+      // this thread goes over all the requests that are currently in flight
+      // If they've closed, it'll close these requests and remove them from the queue
+      g_closing_thread = rsl::thread(wait_for_read_requests);
+    }
+
+    void init(rsl::string_view root)
+    {
+      g_vfs_state_controller.change_state(VfsState::Initializing);
+
+      // Setting the root directory has the effect that all data will be read relative from this directory
+      // This has the same effect as if you would put the working directory to this path
+      set_root(cmdline::get_argument("Root").value_or(root));
+
+      // Start the worker threads, listening to file IO requests and processing them
+      start_threads();
+
+      g_vfs_state_controller.change_state(VfsState::Running);
+
+      REX_LOG(FileSystem, "FileSystem initialized");
+    }
+
+    void set_root(rsl::string_view root)
```

Comment:
INFO: Are we ever required to call this function?
I think we as engine-developers control where the root of the VFS points to and users of the engine should just go with it no?

Reply:
this is more for debugging stuff (eg auto tests)